### PR TITLE
Add Newman and Hamlin method for inverting GRMHD variables

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

To discuss:  
- what is best way of handling multiple recovery methods
- how to generalize EoS to enable Newman-Hamlin for arbitrary EoS

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveRecoveryScheme.hpp
Comment (kidder) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveRecoveryScheme.hpp]: .. -> .

**Review by kidder — COMMENTED** [review]
discussion with @nilsdeppe,  turn enum class into 
namespace PrimitiveRecoveryScheme {
class NewmanHamlin;
}  and extend EoS to have needed functions

**Comment by kidder** [comment]
@ffoucart @teukolsky @nilsdeppe @wthrowe this is ready to look at

File: src/PointwiseFunctions/EquationsOfState/SpecificEnthalpy.hpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/EquationsOfState/SpecificEnthalpy.hpp:9]: This should probably be in the EquationsOfStateGroup

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp
Comment (wthrowe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp:23]: This documentation doesn't render correctly.

File: src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:104]: Most of this documentation doesn't seem to be showing up anywhere, but this problem appears to already exist without this PR.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp
Comment (kidder) [reply src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:104]: yes, any suggestions on how to get it to show up would be appreciated (and as @wthrowe says, the previous documentation doesn't show up either, or at least I couldn't find where it did...)

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp
Comment (kidder) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp:23]: pushed a fixup

File: src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:104]: Since it's not a new issue we don't have to resolve it here.  If anyone has ideas for how to fix it then great, otherwise we can file a bug and move on.

**Review by wthrowe — DISMISSED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp:23]: You have some rendering issues: http://gh.nilsdeppe.com/spectre/pr_901/classgrmhd_1_1ValenciaDivClean_1_1PrimitiveRecoverySchemes_1_1NewmanHamlin.html#details

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp]: `tllde` to `tilde`

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp]: I think you're missing a `}` here

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp]: And what is happening with `\frac` and `/` here?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: Could you make this `UNLIKELY`?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: Since you know this is going to be in general less than max number of iterations (and that's only 50), could you reserve that allocation to avoid us reallocating the vector too often?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.hpp]: More `*`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (teukolsky) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: I think this better be v_squared strictly < 1.0, otherwise next statement will fail.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (teukolsky) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: I'm not sure this is correct. If aitken_residual is not between 0 and 1, no extrapolation is done. In this way, more than 3 values of pressure can end up in the vector. It's the most recent 3 that are used to attempt the extrapolation. In the bitbucket code, this was implemented with a counter.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (teukolsky) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: Is there a reason not to recompute rho and W if we converged on an extrapolation?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (ffoucart) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: In the bitbucket version, this was d_in_cubic<-1.e-12*sqr(momentum_density_dot_magnetic_field). This was needed because roundoff errors occasionally led to just-below-zero results for d_in_cubic. Might be ok to leave it as is until the error appears in Spectre, but maybe leave a comment explaining that getting this error may be ok.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (ffoucart) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: We will probably need the ability to handle v_squared >= 1. here (as in the bitbucket code). If I remember properly, you could end up with a first iteration sending you to an unphysical value of v^2. We then used a slow 1D solver to determine which P would lead to v^2 = 1-epsilon (ideally, epsilon should be chosen to match the maximum Lorentz factor allowed by the atmosphere). Or maybe we drop to a 1D solve when this happens, and don't worry about it.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (ffoucart) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: Yes, we need to handle cases where aitken_pressure.size()>3.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (ffoucart) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: My reading of the code is that we only exit the loop once rho and W have been recomputed (at the beginning of the next iteration through the while loop) - as convergence is not checked by the while loop itself.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: Should this code comment be part of the assert? Not sure, I'll let you decide :)

**Review by nilsdeppe — DISMISSED** [review]
I'm happy for this to be rebased on develop and squashed! :+1:

**Review by ffoucart — DISMISSED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (kidder) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: I'm going to leave it as a comment, as I'll be converting the ASSERT to returning a boost::none in a forthcoming PR

**Review by nilsdeppe — APPROVED** [review]
@teukolsky @ffoucart please approve if you're happy with this or request more changes if not when you have some time :)

**Review by teukolsky — APPROVED** [review]

**Review by ffoucart — APPROVED** [review]

**Comment by nilsdeppe** [comment]
travis failures are timeouts.

**Pull request merged**
